Add email subscription flash partial

sairamya93 opened the pull request:
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What

Adding a partial for email subscribe and unsubscribe flash message

## Why

As part of app consolidation, `email_subscribe_unsubscribe_flash` partial is required for the document type `detailed_guide`. The partial would also be needed for other document types - `call_for_evidence`,`consultation`,`document_collection` and `publication`.

Trello https://trello.com/c/JcPTvK07/564-email-subscription-partial-for-app-consolidation, [Jira issue PNP-6336](https://gov-uk.atlassian.net/browse/PNP-6336)

## How

Based on the code in government-frontend where the apps are currently being rendered.

## Screenshots?

deborahchua committed:
Add GOV.UK Personalisation helper and concern

The helper contains methods used by the `email_subscribe_unsubscribe_flash`
shared partial and was originally introduced in https://github.com/alphagov/government-frontend/pull/2383/commits/2bbe914497ac3097b438328dc85e237dab464861.
It has been renamed to something that feels more suitable and relevant.

We need also to include the GovukPersonalisation controller concern as
this is required to manage GOV.UK accounts, specifically so that we can
use the `account_flash` variable to determine whether the email subscription
banner is shown on a page. See https://github.com/alphagov/govuk_personalisation/blob/main/lib/govuk_personalisation/controller_concern.rb#40

Audit trail:
- https://github.com/alphagov/government-frontend/blob/842a887dc0c83311a9ead2daa08bd430ad0ac347/app/helpers/content_items_helper.rb > app/helpers/govuk_personalisation_helper.rb
- https://github.com/alphagov/government-frontend/blob/842a887dc0c83311a9ead2daa08bd430ad0ac347/app/controllers/content_items_controller.rb#L4 > app/controllers/content_items_controller.rb

deborahchua commented on app/views/shared/_email_subscribe_unsubscribe_flash.html.erb:
```diff
@@ -0,0 +1,31 @@
+<% end %>
+
+<% if show_email_subscription_success_banner?(@account_flash) %>
+  <div class="govuk-grid-row govuk-!-margin-top-3">
```
There have been some recent frontend changes - see https://github.com/alphagov/government-frontend/pull/3655

We will have to update the audit trail link too in the commit message, as the code in that snapshot is now outdated.

leenagupte replied:
+1 to Deb's comment

sairamya93 replied:
Oh thanks! I did have the commit in my local when those changes were made but wanted to push once they got merged in gov-frontend.

sairamya93 committed:
Add local keys for email_subscribe_unsubscribe_flash partial

Ran rake "consolidation:copy_translation[email]"

Audit trail: https://github.com/alphagov/government-frontend/blob/05d961f98ae8fec9ca0c37200aec96ee7bd1d590/config/locales/en.yml#L320

leenagupte commented on spec/helpers/govuk_personalisation_helper_spec.rb:
```diff
@@ -0,0 +1,47 @@
+  describe "#email_subscription_success_banner_heading" do
+    it "displays the subscription success banner when the 'email-subscription-success' flash is present" do
+      account_flash = request.headers["GOVUK-Account-Session"] = GovukPersonalisation::Flash.encode_session("session-id", %w[email-subscription-success])
+      expect(email_subscription_success_banner_heading(account_flash)).to eq("You’ve subscribed to emails about this page")
```
Minor: rather than testing the exact text, you could test that the correct translation has been called. e.g.

expect(email_subscription_success_banner_heading(account_flash)).to eq(I18n.t("email. subscribe_title")

leenagupte commented on app/controllers/content_items_controller.rb:
```diff
@@ -1,4 +1,6 @@
 class ContentItemsController < ApplicationController
+  include GovukPersonalisation::ControllerConcern
+
```
The [set_account_vary_header](https://github.com/alphagov/government-frontend/blob/7c542788cb5c49ec56a1d9af003195428fe4bc59/app/controllers/content_items_controller.rb#L283-L287) method needs to be added to this controller too. Although it doesn't look like it's being called anywhere, what it's actually doing is overriding a method in [GovukPersonalisation::ControllerConcern](https://github.com/alphagov/govuk_personalisation/blob/90da0152c328b9a1b99aa05528c29f2a32fed9ae/lib/govuk_personalisation/controller_concern.rb#L54).

Rather than caching the page by user, the override caches the page by flash message, so that there are only 3 cached versions of the page. For a full explanation see the [commit message](https://github.com/alphagov/government-frontend/commit/8e27396ecfdb6d2ea99cd3abf795402593c8d872) where the override was added.

It would be good to add a request test to check that the `Vary` header has been modified. Something like the original test in [government-frontend](https://github.com/alphagov/government-frontend/blob/7c542788cb5c49ec56a1d9af003195428fe4bc59/test/controllers/content_items_controller_test.rb#L310).

leenagupte reviewed (commented):
This is looking good. 🎉  

I've added some inline comments, most of them are minor.  Modifying the `Vary` header should be added so we continue to cache the pages appropriately.

sairamya93 committed:
Add set_account_vary_header method

 - 'set_account_vary_header' method overrides 'set_account_vary_header' method defined in GovukPersonalisation::ControllerConcern https://github.com/alphagov/govuk_personalisation/blob/90da0152c328b9a1b99aa05528c29f2a32fed9ae/lib/govuk_personalisation/controller_concern.rb#L54 which is called as part of before_action

 - This caches pages based on three flash message variations instead of caching pages per user. The Vary header appends "GOVUK-Account-Session-Exists" and "GOVUK-Account-Session-Flash" to the existing Vary header values as explained here https://github.com/alphagov/government-frontend/commit/8e27396ecfdb6d2ea99cd3abf795402593c8d872

- Audit trail for the method: https://github.com/alphagov/government-frontend/blob/41f451f0bfe1552998ab62ab1ed374f284737f63/app/controllers/content_items_controller.rb#L283
- Audit trail for the test: https://github.com/alphagov/government-frontend/blob/41f451f0bfe1552998ab62ab1ed374f284737f63/test/controllers/content_items_controller_test.rb#L310

leenagupte commented on spec/helpers/govuk_personalisation_helper_spec.rb:
```diff
@@ -3,44 +3,44 @@
     it "displays the subscription success banner when the 'email-subscription-success' flash is present" do
-      account_flash = request.headers["GOVUK-Account-Session"] = GovukPersonalisation::Flash.encode_session("session-id", %w[email-subscription-success])
-      expect(email_subscription_success_banner_heading(account_flash)).to eq("You’ve subscribed to emails about this page")
+      account_flash = GovukPersonalisation::Flash.encode_session("session-id", %w[email-subscription-success])
```
I think something went a bit awry with the rebase. These changes should be fixed up to the previous commit.

sairamya93 replied:
Oh yes I didn't realise that. Thanks I fixed it up now.

leenagupte commented on app/controllers/content_items_controller.rb:
```diff
@@ -46,4 +46,10 @@ def set_locale
     response.headers["X-Frame-Options"] = "DENY"
   end
+
+  def set_account_vary_header
```
Thanks for adding this. And nice audit trail in the commit message! ⭐

leenagupte reviewed (commented):
This is looking great! Thanks for making the requested changes.

There's only one minor comment about the contents of a commit. Once that's sorted out this will be good to go.

sairamya93 committed:
Add email_subscribe_unsubscribe_flash shared partial

It is also used on consultation, detailed guide, document collection
and publication pages.

Commit audit trail:
https://github.com/alphagov/government-frontend/blob/05d961f98ae8fec9ca0c37200aec96ee7bd1d590/app/views/shared/_email_subscribe_unsubscribe_flash.html.erb

leenagupte reviewed (approved):
👍
